devito/types/basic.py

```diff
@@ -879,6 +880,12 @@ def __new__(cls, *args, **kwargs):
             # let's just return `function` itself
             return function
 
+        # If dimensions have been replaced, then it is necessary to remove halo
+        # and padding kwargs so that they are rebuilt with the new dimensions
+        if function is not None and function.dimensions != dimensions:
```

Comment:
That should not be a valid combination. If `function` is not None it *has* to have the same dimensions

Reply:
this is just for reconstruction; see one of the latest threads on slack in #development